src/parse.d

```diff
@@ -3070,97 +3068,116 @@ public:
-                    if (tpl)
-                    {
-                        Expression tempCons = parseConstraint();
-                        if (tempCons)
-                        {
-                            if (constraint)
-                                error("members expected");
-                            else
-                                constraint = tempCons;
-                        }
-                    }
-                    if (token.value != TOKlcurly)
-                        error("members expected");
-                }
-
-                if (tok == TOKclass)
-                {
-                    bool inObject = md && !md.packages && md.id == Id.object;
-                    a = new ClassDeclaration(loc, id, baseclasses, inObject);
-                }
-                else
-                    a = new InterfaceDeclaration(loc, id, baseclasses);
-                break;
-            }
-        case TOKstruct:
-            if (id)
-                a = new StructDeclaration(loc, id);
-            else
-                anon = 1;
-            break;
-
-        case TOKunion:
-            if (id)
-                a = new UnionDeclaration(loc, id);
-            else
-                anon = 2;
-            break;
+            nextToken();
+            baseclasses = parseBaseClasses();
+        }
 
-        default:
-            assert(0);
+        if (token.value == TOKif)
+        {
+            if (constraint)
+                error("multiple constraints not allowed");
+            constraint = parseConstraint();
         }
-        if (a && token.value == TOKsemicolon)
+        if (constraint)
         {
-            nextToken();
+            if (!id)
+                error("template constraints not allowed for anonymous %s", Token.toChars(tok));
+            if (!tpl)
+                error("template constraints only allowed for templates");
         }
-        else if (token.value == TOKlcurly)
+
+        Dsymbols* members = null;
+        if (token.value == TOKlcurly)
         {
+            //printf("aggregate definition\n");
             const lookingForElseSave = lookingForElse;
             lookingForElse = Loc();
-            //printf("aggregate definition\n");
             nextToken();
-            Dsymbols* decl = parseDeclDefs(0);
+            members = parseDeclDefs(0);
             lookingForElse = lookingForElseSave;
             if (token.value != TOKrcurly)
+            {
+                /* { */
                 error("} expected following members in %s declaration at %s",
                     Token.toChars(tok), loc.toChars());
+            }
+            nextToken();
+        }
+        else if (token.value == TOKsemicolon && id)
+        {
+            if (baseclasses || constraint)
+                error("members expected");
             nextToken();
-            if (anon)
+        }
+        else
+        {
+            error("{ } expected following %s declaration", Token.toChars(tok));
```

Comment:
I think the message would a bit more user friendly if it includes the unexpected token, e.g.:
```
"{ } expected following %s declaration, not %s"
```

Is this something easy to do?